feat: add Repo Activity Racing Bar

andyhuang18 opened the pull request:
## Brief Information

This pull request is in the type of ([more info](https://github.com/angular/angular/blob/main/CONTRIBUTING.md#type) about types):

- [ ] build
- [ ] ci
- [ ] docs
- [x] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] test

Related issues ([all available keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

- resolve #687 

## Details
<!-- What did you do in this PR?  -->
I have laid a foundation for this feature, which can be viewed in the option list. But the actual functionality has not yet been implemented. I'll follow up as soon as I have time!

![image](https://github.com/hypertrons/hypertrons-crx/assets/50283262/88cffdca-e5bd-4d54-8720-51c9bb340350)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/hypertrons/hypertrons-crx/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/hypertrons/hypertrons-crx)

## Others
<!-- Other information you want to share.  -->
<!-- If none to share, leave an "N.A."  -->

andyhuang18 commented:
The function of the main body has been initially realized, and the following is a demonstration video. I have not modified the text on the right.


https://github.com/hypertrons/hypertrons-crx/assets/50283262/6ad5e666-68cd-4343-a806-3b0d30872cac

tyn1998 commented:
Hey @andyhuang18, the animation in your demo is very laggy, do you know why?

andyhuang18 commented:
I changed the `updateFrequency` to 4000. Maybe `updateFrequency` should be adjusted to an appropriate value.

andyhuang18 commented:
I have successfully completed the development of this feature, and the `replay` function is also available. In the future, other new functions can be added, such as positioning time, etc. 😄

wxharry commented:
> I have successfully completed the development of this feature, and the `replay` function is also available. In the future, other new functions can be added, such as positioning time, etc. 😄

Nicely done! I just tested your work ant it is amazing! It moves smoothly and looks awesome. Here is some advice on this:
1. style change. To make style identical, would you mind wrapping up chart with a `div` container with `hypertrons-crx-border` like the other two graphs in the `perceptor` tab.
2. layout change. The `Period: 90 days` is meaningless here. I think we could move that `replay` button to that place and also change the style of the button a little to make it more natural. What do you say?

Again, thanks for your incredible work! 👍

tyn1998 commented:
Hey @andyhuang18, thanks for your hard work!

### one
I pulled your code and tested it with my hands, finding that the feature occasionally cannot be displayed with the following error output in the console:

<img width="1512" alt="image" src="https://github.com/hypertrons/hypertrons-crx/assets/32434520/5bdf811c-74a2-4286-822d-0159bef55294">

I have not found a way to replicate the bug yet. But I guess it is possibly related to feature loading timing.


> As for the position of `replay`,I think it would be reasonable to put the `replay` button below the chart, and maybe add other buttons to manipulate the chart in the future~

### two
Please give the Replay button a new home:

<img width="1278" alt="image" src="https://github.com/hypertrons/hypertrons-crx/assets/32434520/486d1580-8769-416f-a896-cbd128ea74a5">

The displacement of the button may require you to learn and use more skills in HTML, CSS, React. Good luck :)

### three
I have not read the code changes yet. I assume they will be reviewed in days.

wxharry commented:
Hi @andyhuang18, just change the to "feat: add Repo Activity Racing Bar". You are not just implementing this, you are adding the entire feature in this pr. 😄

wxharry commented:
Hi, @andyhuang18 . There are two style changes required for this PR:
1. add a 15px padding to the border or a 15px margin to the racing bar. This is because some username is long and the name is at the edge
2. move the replay button to the top-right corner. I understand your concern about adding more features, but we should move the button for now. The top-right corner is a place for user interaction which is used in the insight tab at Github.

Thanks for your contributions and feel free to leave comments or dm me~

andyhuang18 commented:
OK! Thanks harry for the suggestion! I am modifying my code. Commit can be done today. 😆

andyhuang18 commented:
Hi~ @tyn1998 @wxharry I've done all the development of this feature so far. 
A brief description of the last few updates：
- Moved the Replay button to the upper right corner of the interface, and modified the style to match the GitHub button.

![image](https://github.com/hypertrons/hypertrons-crx/assets/50283262/0f829669-49b5-4b74-8596-ee3debc2d634)
- Provides `i18n` support.
- Since I am using the `webstorm` editor, I updated gitignore. In the future, updates in the `.idea` folder will be ignored by git.

PS: I didn't know enough about the update mode of echart before, so `[ECharts] Instance ec_1691746561518 has been disposed` will always be prompted in the console of the browser during replay. At present, this problem has been solved, and the reason is: because I achieved the purpose of replay by re-creating the echart instance, the first instance has been dispose after re-creation, and the second hook function is still updating the data of the first instance chart , so an error will be reported. The solution is to first determine whether an instance exists when updating data.

If you have any additional suggestions, please let me know, thank you~

wxharry commented:
> Hi~ @tyn1998 @wxharry I've done all the development of this feature so far. A brief description of the last few updates：
>  ... 
> If you have any additional suggestions, please let me know, thank you~

Hey! Thanks again for your contributions. This feature is really awesome! I have reviewed your code and left a comment. It won't take long to resolve and I think this feature is ready to merge. I am so excited to see the racing bar released.

andyhuang18 commented:
> > Hi~ @tyn1998 @wxharry I've done all the development of this feature so far. A brief description of the last few updates：
> > ...
> > If you have any additional suggestions, please let me know, thank you~
> 
> Hey! Thanks again for your contributions. This feature is really awesome! I have reviewed your code and left a comment. It won't take long to resolve and I think this feature is ready to merge. I am so excited to see the racing bar released.

Hi @wxharry. Where is the problem? Seems like I didn't see your comment.  🤔

tyn1998 commented on src/pages/ContentScripts/features/repo-activity-racing-bar/RacingBar.tsx:
```diff
@@ -0,0 +1,201 @@
+      if (!instance.isDisposed()) {
+        instance.dispose();
+      }
+      // clear timer if user replay the chart before it finishes
+      if (timer) {
+        clearTimeout(timer);
+      }
```
Timer should be cleared.

tyn1998 commented on src/pages/ContentScripts/features/repo-activity-racing-bar/RacingBar.tsx:
```diff
@@ -0,0 +1,201 @@
+  const months = Object.keys(data);
+  let i = 0;
+
+  const playNext = () => {
+    updateMonth(instance, data, months[i]);
+    i++;
+    if (i < months.length) {
+      timer = setTimeout(playNext, updateFrequency);
+    }
+  };
+
+  playNext();
```
another way to play

tyn1998 commented on src/pages/ContentScripts/features/repo-activity-racing-bar/RacingBar.tsx:
```diff
@@ -0,0 +1,201 @@
+      height: 20,
+    };
+  });
+  // @ts-ignore
+  option.yAxis.axisLabel.rich = rich;
+  // @ts-ignore
+  option.series[0].data = data[month];
+  // @ts-ignore
+  option.graphic.elements[0].style.text = month;
```
`// @ts-ignore` is like `xxx: any`, we should fix them all one day :D

tyn1998 commented on src/pages/ContentScripts/features/repo-activity-racing-bar/RacingBar.tsx:
```diff
@@ -0,0 +1,201 @@
+  // @ts-ignore
+  option.graphic.elements[0].style.text = month;
+
+  // it seems that hidden bars are also rendered, so when each setOption merge more data into the chart,
+  // the fps goes down. So we use notMerge to avoid merging data. But this disables the xAxis animation.
+  // Hope we can find a better solution.
+  instance.setOption(option, {
+    notMerge: true,
+  });
+};
```
microsoft/vscode has a lot of contributors, which slows down the animation when played at ~2018 or so.

![image](https://github.com/hypertrons/hypertrons-crx/assets/32434520/604c96c3-b2d8-4648-929c-06b6ee7bf1bb)

`notMerge: true` is the best resolution I can give for now.

tyn1998 reviewed (approved):
Hi @andyhuang18, I have pushed several commits to improve the code.

Your contribution is great, thank you very much!

tyn1998 commented:
Hi @wxharry, please take some time to review my recent commits to this branch. If you think the PR is ready to go, just `/approve` it :)

tyn1998 commented:
> Seems like I didn't see your comment. 🤔

+1

tyn1998 commented on src/pages/ContentScripts/features/repo-activity-racing-bar/RacingBar.tsx:
```diff
@@ -0,0 +1,165 @@
+      })(i);
+    }
+
+    // @ts-ignore
```
Yeah, `@ts-ignore` and `any` should be avoided with best effort. In my commits, I removed some of them, but still left some others. My time is limited🤣 .

wxharry replied:
We can have Andy fixing this later as a patch. I will merge this feature anyway but you can still improve your work. @andyhuang18 What do you say?

andyhuang18 replied:
OK

wxharry commented:
There is an inconsistence when open perceptor tab in different repos. For our repo, the racing bar is the first section, but in repos like vscode, it is the last one. It is not strongly related to this feature, so I will create a new issue. This feature looks good to me. Thanks @andyhuang18  @tyn1998 for your contributions!

/approve
